app/main.py

```diff
@@ -1,6 +1,23 @@
-from typing import Callable
+from typing import Callable, Any
 
 
 def cache(func: Callable) -> Callable:
-    # Write your code here
-    pass
+    cached_data = {}
+    immutable_type = (int, float, str, tuple, bool, type(None))
+
+    def argument_based_cacher(*args) -> Any:
+        nonlocal cached_data
+        if all([isinstance(variable, immutable_type) for variable in args]):
+            func_key = tuple(args)
```

Comment:
args is already a tuple, and there's no need to convert it again using tuple(args).